File: app/src/main/java/com/orgzly/android/ui/fragments/SettingsFragment.java
package com.orgzly.android.ui.fragments;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.os.Build;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.Preference;
import android.support.annotation.StringRes;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ListView;

import com.github.machinarius.preferencefragment.PreferenceFragment;
import com.orgzly.BuildConfig;
import com.orgzly.R;
import com.orgzly.android.Shelf;
import com.orgzly.android.Notifications;
import com.orgzly.android.prefs.AppPreferences;
import com.orgzly.android.prefs.ListPreferenceWithValueAsSummary;
import com.orgzly.android.provider.clients.ReposClient;
import com.orgzly.android.repos.Repo;
import com.orgzly.android.ui.FragmentListener;
import com.orgzly.android.ui.NoteStateSpinner;
import com.orgzly.android.ui.util.ActivityUtils;
import com.orgzly.android.util.LogUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;

/**
 * Displays settings.
 */
public class SettingsFragment extends PreferenceFragment
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    public static final String TAG = SettingsFragment.class.getName();

    public static final String FRAGMENT_TAG = SettingsFragment.class.getName();

    private static final @StringRes int[] REQUIRE_ACTIVITY_RESTART = new int [] {
            R.string.pref_key_font_size,
            R.string.pref_key_color_scheme,
            R.string.pref_key_layout_direction
    };

    private Preference mReposPreference;
    private SettingsFragmentListener mListener;
    public static SettingsFragment getInstance() {
        return new SettingsFragment();
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.preferences);

        mReposPreference = findPreference(getString(R.string.pref_key_repos));

        findPreference(getString(R.string.pref_key_clear_database))
                .setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        if (mListener != null) {
                            mListener.onDatabaseClearRequest();
                        }
                        return true;
                    }
                });
        findPreference(getString(R.string.pref_key_reload_getting_started))
                .setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                    @Override
                    public boolean onPreferenceClick(Preference preference) {
                        if (mListener != null) {
                            mListener.onGettingStartedNotebookReloadRequest();
                        }
                        return true;
                    }
                });

        setupVersionPreference();

        /* Receive onCreateOptionsMenu() call, to remove search menu item. */
        setHasOptionsMenu(true);

        setDefaultStateForNewNote();

        /* Disable preference for changing the layout, if not on API version that supports that. */
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR1) {
            Preference layoutDirectionPref = findPreference(getString(R.string.pref_key_layout_direction));
            layoutDirectionPref.setEnabled(false);
        }

        /* Update preferences which depend on multiple others. */
        updateRemindersPreferences();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = super.onCreateView(inflater, container, savedInstanceState);

        /*
         * Set fragment's background.
         */
        if (view != null) {
            int[] textSizeAttr = new int[] { R.attr.item_book_card_bg_color};
            TypedArray typedArray = view.getContext().obtainStyledAttributes(textSizeAttr);
            int color = typedArray.getColor(0, -1);
            typedArray.recycle();

            if (color != -1) {
                view.setBackgroundColor(color);
            }
        }

        /*
         * Add some padding to list.
         */
        if (view != null) {
            ListView list = (ListView) view.findViewById(android.R.id.list);
            if (list != null) {
                int h = (int) getResources().getDimension(R.dimen.fragment_horizontal_padding);
                int v = (int) getResources().getDimension(R.dimen.fragment_vertical_padding);
                list.setPadding(h, v, h, v);
            }
        }

        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        mListener = (SettingsFragmentListener) getActivity();
    }

    @Override
    public void onResume() {
        super.onResume();

        updateUserReposPreferenceSummary();

        /* Start to listen for any preference changes. */
        android.preference.PreferenceManager.getDefaultSharedPreferences(getActivity())
                .registerOnSharedPreferenceChangeListener(this);


        announceChangesToActivity();
    }

    private void announceChangesToActivity() {
        if (mListener != null) {
            mListener.announceChanges(
                    SettingsFragment.FRAGMENT_TAG,
                    getString(R.string.fragment_settings_title),
                    null,
                    0);
        }
    }

    @Override
    public void onPause() {
        super.onPause();

        /* Stop listening for preference changed. */
        android.preference.PreferenceManager.getDefaultSharedPreferences(getActivity())
                .unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onDetach() {
        super.onDetach();

        mListener = null;
    }

    private void setupVersionPreference() {
        Preference versionPreference = findPreference(getString(R.string.pref_key_version));

        /* Set summary to the current version string, appending suffix for the flavor. */
        versionPreference.setSummary(BuildConfig.VERSION_NAME + BuildConfig.VERSION_NAME_SUFFIX);

        /* Display changelog dialog when version is clicked. */
        versionPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                if (mListener != null) {
                    mListener.onWhatsNewDisplayRequest();
                }

                return true;
            }
        });

    }

    /**
     * Updates preference's summary with a list of configured user repos.
     */
    public void updateUserReposPreferenceSummary() {
        Map<String, Repo> repos = ReposClient.getAll(getActivity().getApplicationContext());

        if (repos.isEmpty()) {
            mReposPreference.setSummary(R.string.no_user_repos_configured_pref_summary);

        } else {
            List<String> list = new ArrayList<>();
            for (Repo repo : repos.values()) {
                list.add(repo.toString());
            }
            Collections.sort(list);
            mReposPreference.setSummary(TextUtils.join("\n", list));
        }
    }

    /**
     * Callback for options menu.
     */
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        /*
         * Clear the menu. We want new SettingsActivity anyway,
         * to remove as much code from MainActivity as possible.
         */
        menu.clear();
    }

    /**
     * Called when a shared preference is modified in any way.
     * Used to update AppPreferences' static values and do any required post-settings-change work.
     */
    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, sharedPreferences, key);

        Activity activity = getActivity();

        if (activity == null) {
            return;
        }

        /* State keywords. */
        if (getString(R.string.pref_key_states).equals(key)) {
            AppPreferences.updateStaticKeywords(getContext());

            /* Re-parse notes. */
            ActivityUtils.closeSoftKeyboard(activity);

            if (mListener != null) {
                mListener.onStateKeywordsPreferenceChanged();
            }

            setDefaultStateForNewNote();
        }

        /* Recreate activity if preference change requires it. */
        for (int res: REQUIRE_ACTIVITY_RESTART) {
            if (key.equals(getString(res))) {
                activity.recreate();
                break;

            }
        }

        /* Cancel or create an new-note ongoing notification. */
        if (getString(R.string.pref_key_new_note_notification).equals(key)) {
            if (AppPreferences.newNoteNotification(getContext())) {
                Notifications.createNewNoteNotification(getContext());
            } else {
                Notifications.cancelNewNoteNotification(getContext());
            }
        }

        /* Update default priority when minimum priority changes. */
        if (getString(R.string.pref_key_min_priority).equals(key)) {

            String defPri = AppPreferences.defaultPriority(getContext());
            String minPri = sharedPreferences.getString(key, null);

            // Default priority is lower then minimum
            if (defPri.compareToIgnoreCase(minPri) > 0) { // minPri -> defPri
                /* Must use preference directly to update the view too. */
                ListPreferenceWithValueAsSummary pref =
                        (ListPreferenceWithValueAsSummary)
                                findPreference(getString(R.string.pref_key_default_priority));
                pref.setValue(minPri);
            }
        }

        /* Update minimum priority when default priority changes. */
        if (getString(R.string.pref_key_default_priority).equals(key)) {
            String minPri = AppPreferences.minPriority(getContext());
            String defPri = sharedPreferences.getString(key, null);

            // Default priority is lower then minimum
            if (minPri.compareToIgnoreCase(defPri) < 0) { // minPri -> defPri
                /* Must use preference directly to update the view too. */
                ListPreferenceWithValueAsSummary pref =
                        (ListPreferenceWithValueAsSummary)
                                findPreference(getString(R.string.pref_key_min_priority));
                pref.setValue(defPri);
            }
        }

        if (getString(R.string.pref_key_use_reminders_for_scheduled_times).equals(key)) {
            /* Reset last run time if reminders are being enabled or disabled. */
            AppPreferences.reminderServiceLastRun(getContext(), 0L);
        }

        if (getString(R.string.pref_key_use_reminders_for_scheduled_times).equals(key)) {
            updateRemindersPreferences();
        }


        /* Always notify if settings are modified.
         * Changing states or priorities can affect the displayed data for example.
         */
        Shelf.notifyDataChanged(getContext());
    }

    private void updateRemindersPreferences() {
        boolean remindersEnabled = ((CheckBoxPreference) findPreference(getString(R.string.pref_key_use_reminders_for_scheduled_times))).isChecked();

        findPreference(getString(R.string.pref_key_reminders_sound)).setEnabled(remindersEnabled);
        findPreference(getString(R.string.pref_key_reminders_vibrate)).setEnabled(remindersEnabled);
    }

    /**
     * Update list of possible states that can be used as default for a new note.
     */
    private void setDefaultStateForNewNote() {
        /* NOTE followed by to-do keywords */
        LinkedHashSet<CharSequence> entries = new LinkedHashSet<>();
        entries.add(NoteStateSpinner.NO_STATE_KEYWORD);
        entries.addAll(AppPreferences.todoKeywordsSet(getContext()));
        CharSequence[] entriesArray = entries.toArray(new CharSequence[entries.size()]);

        /* Set possible values. */
        ListPreferenceWithValueAsSummary pref = (ListPreferenceWithValueAsSummary) findPreference(getString(R.string.pref_key_new_note_state));
        pref.setEntries(entriesArray);
        pref.setEntryValues(entriesArray);

        /* Set current value. */
        String value = AppPreferences.newNoteState(getContext());
        if (entries.contains(value)) {
            pref.setValue(value);
        } else {
            pref.setValue(NoteStateSpinner.NO_STATE_KEYWORD);
        }
    }

    public interface SettingsFragmentListener extends FragmentListener {
        void onStateKeywordsPreferenceChanged();

        void onDatabaseClearRequest();

        void onGettingStartedNotebookReloadRequest();

        void onWhatsNewDisplayRequest();
    }
}

File: app/src/main/java/com/orgzly/android/ui/settings/SettingsFragment.kt
package com.orgzly.android.ui.settings

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.os.Handler
import androidx.annotation.StringRes
import androidx.preference.*
import com.orgzly.BuildConfig
import com.orgzly.R
import com.orgzly.android.AppIntent
import com.orgzly.android.SharingShortcutsManager
import com.orgzly.android.prefs.*
import com.orgzly.android.reminders.RemindersScheduler
import com.orgzly.android.ui.CommonActivity
import com.orgzly.android.ui.NoteStates
import com.orgzly.android.ui.notifications.Notifications
import com.orgzly.android.ui.util.KeyboardUtils
import com.orgzly.android.usecase.NoteReparseStateAndTitles
import com.orgzly.android.usecase.NoteSyncCreatedAtTimeWithProperty
import com.orgzly.android.usecase.UseCase
import com.orgzly.android.util.AppPermissions
import com.orgzly.android.util.LogUtils
import com.orgzly.android.widgets.ListWidgetProvider

/**
 * Displays settings.
 */
class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {
    private var listener: Listener? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)

        listener = activity as Listener
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource()

        setupPreferences()
    }

    private fun addPreferencesFromResource() {
        when (val resourceName = arguments?.getString(ARG_RESOURCE)) {
            null -> addPreferencesFromResource(R.xml.prefs) // Headings
            in PREFS_RESOURCES -> addPreferencesFromResource(PREFS_RESOURCES.getValue(resourceName))
        }
    }

    private fun setupPreferences() {
        /* Make icons lighter. */
//        preferenceScreen?.let {
//            for (i in 0 until it.preferenceCount) {
//                it.getPreference(i)?.icon?.alpha = 140
//            }
//        }


        preference(R.string.pref_key_clear_database)?.let {
            it.setOnPreferenceClickListener {
                listener?.onDatabaseClearRequest()
                true
            }
        }

        preference(R.string.pref_key_reload_getting_started)?.let {
            it.setOnPreferenceClickListener {
                listener?.onGettingStartedNotebookReloadRequest()
                true
            }
        }

        setupVersionPreference()

        setDefaultStateForNewNote()

        preference(R.string.pref_key_file_absolute_root)?.let {
            val pref = it as EditTextPreference
            pref.text = AppPreferences.fileAbsoluteRoot(context)
        }

        preference(R.string.pref_key_file_relative_root)?.let {
            val pref = it as EditTextPreference
            pref.text = AppPreferences.fileRelativeRoot(context)
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            preference(R.string.pref_key_reminders_notification_settings_V26)?.let {
                preferenceScreen.removePreference(it)
            }
        } else {
            preference(R.string.pref_key_reminders_notification_settings_preV26)?.let {
                preferenceScreen.removePreference(it)
            }
        }

        /* Update preferences which depend on multiple others. */
        updateRemindersScreen()
    }

    private fun setupVersionPreference() {
        preference(R.string.pref_key_version)?.let { pref ->
            /* Set summary to the current version string, appending suffix for the flavor. */
            pref.summary = BuildConfig.VERSION_NAME + BuildConfig.VERSION_NAME_SUFFIX

            /* Display changelog dialog when version is clicked. */
            pref.setOnPreferenceClickListener {
                listener?.onWhatsNewDisplayRequest()
                true
            }
        }
    }

    /*
     * Display custom preference's dialog.
     */
    override fun onDisplayPreferenceDialog(preference: Preference) {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, preference)

        when (preference) {
            is StatesPreference ->
                displayCustomPreferenceDialogFragment(
                        StatesPreferenceFragment.getInstance(preference),
                        StatesPreferenceFragment.FRAGMENT_TAG)

            is IntegerPreference ->
                displayCustomPreferenceDialogFragment(
                        IntegerPreferenceFragment.getInstance(preference),
                        IntegerPreferenceFragment.FRAGMENT_TAG)

            is TimePreference ->
                displayCustomPreferenceDialogFragment(
                        TimePreferenceFragment.getInstance(preference),
                        TimePreferenceFragment.FRAGMENT_TAG)

            is NotePopupPreference ->
                displayCustomPreferenceDialogFragment(
                    NotePopupPreferenceFragment.getInstance(preference),
                    NotePopupPreferenceFragment.FRAGMENT_TAG)


            else -> super.onDisplayPreferenceDialog(preference)
        }
    }

    private fun displayCustomPreferenceDialogFragment(
            fragment: PreferenceDialogFragmentCompat,
            tag: String
    ) {
        fragmentManager?.let {
            fragment.setTargetFragment(this, 0)
            fragment.show(it, tag)
        }
    }

    override fun onResume() {
        super.onResume()

        listener?.onTitleChange(preferenceScreen?.title)

        /* Start to listen for any preference changes. */
        PreferenceManager.getDefaultSharedPreferences(requireActivity())
                .registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()

        /* Stop listening for preference changed. */
        PreferenceManager.getDefaultSharedPreferences(requireActivity())
                .unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onDetach() {
        super.onDetach()

        listener = null
    }

    /**
     * Called when a shared preference is modified in any way.
     * Used to update AppPreferences' static values and do any required post-settings-change work.
     */
    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, sharedPreferences, key)

        val activity = activity as? CommonActivity ?: return

        when (key) {
            getString(R.string.pref_key_note_popup_buttons_in_book_left),
            getString(R.string.pref_key_note_popup_buttons_in_book_right),
            getString(R.string.pref_key_note_popup_buttons_in_query_left),
            getString(R.string.pref_key_note_popup_buttons_in_query_right) -> {
                findPreference<NotePopupPreference>(key)?.summary =
                    sharedPreferences.getString(key, "")

            }

            // State keywords
            getString(R.string.pref_key_states) -> {
                AppPreferences.updateStaticKeywords(context)

                /* Re-parse notes. */
                KeyboardUtils.closeSoftKeyboard(activity)
                listener?.onNotesUpdateRequest(NoteReparseStateAndTitles())

                setDefaultStateForNewNote()
            }

            // Created-at property
            getString(R.string.pref_key_is_created_at_added),
            getString(R.string.pref_key_created_at_property) -> {
                if (AppPreferences.createdAt(context)) {
                    listener?.onNotesUpdateRequest(NoteSyncCreatedAtTimeWithProperty())
                }
            }

            // Cancel or create an new-note ongoing notification
            getString(R.string.pref_key_ongoing_notification),
            getString(R.string.pref_key_ongoing_notification_priority) -> {
                if (AppPreferences.newNoteNotification(context)) {
                    Notifications.showOngoingNotification(context)
                } else {
                    Notifications.cancelNewNoteNotification(context)
                }
            }

            // Update default priority when minimum priority changes
            getString(R.string.pref_key_min_priority) -> {
                preference(R.string.pref_key_default_priority)?.let {
                    val pref = it as ListPreference

                    val defPri = AppPreferences.defaultPriority(context)
                    val minPri = sharedPreferences.getString(key, null)

                    // Default priority is lower then minimum
                    if (minPri != null && defPri.compareTo(minPri, ignoreCase = true) > 0) {
                        // minPri -> defPri
                        // Must use preference directly to update the view too
                        pref.value = minPri
                    }
                }
            }

            // Update minimum priority when default priority changes
            getString(R.string.pref_key_default_priority) -> {
                preference(R.string.pref_key_min_priority)?.let {
                    val pref = it as ListPreference

                    val minPri = AppPreferences.minPriority(context)
                    val defPri = sharedPreferences.getString(key, null)

                    // Default priority is lower then minimum
                    if (defPri != null && minPri.compareTo(defPri, ignoreCase = true) < 0) {
                        // minPri -> defPri
                        // Must use preference directly to update the view too
                        pref.value = defPri
                    }
                }
            }

            // Update widget for changed style
            getString(R.string.pref_key_widget_color_scheme),
            getString(R.string.pref_key_widget_opacity),
            getString(R.string.pref_key_widget_font_size),
            getString(R.string.pref_key_widget_display_checkmarks),
            getString(R.string.pref_key_widget_display_book_name),
            getString(R.string.pref_key_widget_update_frequency) -> {
                val intent = Intent(context, ListWidgetProvider::class.java).apply {
                    action = AppIntent.ACTION_UPDATE_LAYOUT_LIST_WIDGET
                }
                context?.sendBroadcast(intent)
            }

            // Reminders for scheduled notes - reset last run time
            getString(R.string.pref_key_use_reminders_for_scheduled_times) ->
                AppPreferences.reminderLastRunForScheduled(context, 0L)

            // Reminders for deadlines - reset last run time
            getString(R.string.pref_key_use_reminders_for_deadline_times) ->
                AppPreferences.reminderLastRunForDeadline(context, 0L)

            // Reminders for events - reset last run time
            getString(R.string.pref_key_use_reminders_for_event_times) ->
                AppPreferences.reminderLastRunForEvents(context, 0L)

            // Display images inline enabled - request permission
            getString(R.string.pref_key_images_enabled) -> {
                if (AppPreferences.imagesEnabled(context)) {
                    Handler().post {
                        AppPermissions.isGrantedOrRequest(
                                activity, AppPermissions.Usage.EXTERNAL_FILES_ACCESS)
                    }
                }
            }
        }

        updateRemindersScreen()

        /* Always notify about possibly changed data, if settings are modified.
         *
         * For example:
         * - Changing states or priorities can affect the displayed data
         * - Enabling or disabling reminders needs to trigger reminder service notification
         */
        RemindersScheduler.notifyDataSetChanged(requireContext())
        ListWidgetProvider.notifyDataSetChanged(requireContext())
        SharingShortcutsManager().replaceDynamicShortcuts(requireContext())
    }

    private fun updateRemindersScreen() {
        val scheduled = preference(R.string.pref_key_use_reminders_for_scheduled_times)
        val deadline = preference(R.string.pref_key_use_reminders_for_deadline_times)
        val event = preference(R.string.pref_key_use_reminders_for_event_times)

        if (scheduled != null && deadline != null && event != null) {
            val remindersEnabled =
                    (scheduled as TwoStatePreference).isChecked
                            || (deadline as TwoStatePreference).isChecked
                            || (event as TwoStatePreference).isChecked

            /* These do not exist on Oreo and later */
            preference(R.string.pref_key_reminders_sound)?.isEnabled = remindersEnabled
            preference(R.string.pref_key_reminders_led)?.isEnabled = remindersEnabled
            preference(R.string.pref_key_reminders_vibrate)?.isEnabled = remindersEnabled

            /* This does not exist before Oreo. */
            preference(R.string.pref_key_reminders_channel_notification_settings)?.isEnabled = remindersEnabled

            preference(R.string.pref_key_snooze_time)?.isEnabled = remindersEnabled
            preference(R.string.pref_key_snooze_type)?.isEnabled = remindersEnabled
            preference(R.string.pref_key_daily_reminder_time)?.isEnabled = remindersEnabled
        }
    }

    /**
     * Update list of possible states that can be used as default for a new note.
     */
    private fun setDefaultStateForNewNote() {
        preference(R.string.pref_key_new_note_state)?.let {
            val pref = it as ListPreference

            /* NOTE followed by to-do keywords */
            val entries = LinkedHashSet<CharSequence>()
            entries.add(NoteStates.NO_STATE_KEYWORD)
            entries.addAll(AppPreferences.todoKeywordsSet(context))
            val entriesArray = entries.toTypedArray()

            /* Set possible values. */
            pref.entries = entriesArray
            pref.entryValues = entriesArray

            /* Set current value. */
            val value = AppPreferences.newNoteState(context)
            if (entries.contains(value)) {
                pref.value = value
            } else {
                pref.value = NoteStates.NO_STATE_KEYWORD
            }
        }
    }

    override fun onPreferenceTreeClick(preference: Preference): Boolean {
        if (preference is PreferenceScreen) {
            preference.key?.let { key ->
                if (key in PREFS_RESOURCES) {
                    listener?.onPreferenceScreen(key)
                    return true
                }
            }
        }

        return super.onPreferenceTreeClick(preference)
    }

    private fun preference(@StringRes resId: Int): Preference? {
        return findPreference(getString(resId))
    }

    interface Listener {
        fun onNotesUpdateRequest(action: UseCase)
        fun onDatabaseClearRequest()
        fun onGettingStartedNotebookReloadRequest()
        fun onWhatsNewDisplayRequest()
        fun onPreferenceScreen(resource: String)
        fun onTitleChange(title: CharSequence?)
    }

    companion object {
        private val TAG: String = SettingsFragment::class.java.name

        val FRAGMENT_TAG: String = SettingsFragment::class.java.name

        private const val ARG_RESOURCE = "resource"

        /* Using headers file & fragments didn't work well - transitions were
         * not smooth, previous fragment would be briefly displayed.
         */
        @JvmField
        val PREFS_RESOURCES: HashMap<String, Int> = hashMapOf(
                "prefs_screen_look_and_feel" to R.xml.prefs_screen_look_and_feel,
                "prefs_screen_notebooks" to R.xml.prefs_screen_notebooks,
                "prefs_screen_notifications" to R.xml.prefs_screen_notifications,
                "prefs_screen_reminders" to R.xml.prefs_screen_reminders,
                "prefs_screen_sync" to R.xml.prefs_screen_sync,
                "prefs_screen_auto_sync" to R.xml.prefs_screen_auto_sync, // Sub-screen
                "prefs_screen_org_file_format" to R.xml.prefs_screen_org_file_format, // Sub-screen
                "prefs_screen_org_mode_tags_indent" to R.xml.prefs_screen_org_mode_tags_indent, // Sub-screen
                "prefs_screen_widget" to R.xml.prefs_screen_widget, // Sub-screen
                "prefs_screen_developer" to R.xml.prefs_screen_developer, // Sub-screen
                "prefs_screen_app" to R.xml.prefs_screen_app
        )

        fun getInstance(res: String? = null): SettingsFragment {
            val fragment = SettingsFragment()

            val args = Bundle()
            args.putString(ARG_RESOURCE, res)
            fragment.arguments = args

            return fragment
        }
    }
}
